data-pipeline/events-flattener/src/main/java/org/ekstep/ep/samza/service/EventsFlattenerService.java

```diff
@@ -0,0 +1,112 @@
+
+    /**
+     * Constructor of the EventFlattenService
+     *
+     * @param config - Configurations of the EventsFlatten Samza job.
+     */
+    public EventsFlattenerService(EventsFlattenerConfig config) {
+        this.config = config;
+    }
+
+    /**
+     * @param source
+     * @param sink
+     */
+    public void process(EventsFlattenerSource source, EventsFlattenerSink sink) {
+        Event event = source.getEvent();
+        if (event.eid().equals(config.getFlattenEventName())) {
+            try {
+                this.toFlatten(event, getClonedEventObject(event), sink);
+            } catch (Exception e) {
+                e.printStackTrace();
+                LOGGER.error(null,
+                        format("EXCEPTION. PASSING EVENT THROUGH AND ADDING IT TO EXCEPTION TOPIC. EVENT: {0}, EXCEPTION:",
+                                event),
+                        e);
+                sink.toErrorTopic(event, e.getMessage());
+            }
+        } else {
+            sink.toSuccessTopic(event);
+        }
+    }
+
+    /**
+     * Method to flatten the SHARE EVENT Object to multiple share event.
+     *
+     * @param orginalEvent - SHARE Event object.
+     * @param sink         - Object to push the event to kafka sink
+     */
+    private void toFlatten(Event orginalEvent, Event clonedEvent, EventsFlattenerSink sink) {
+        String objectId = null;
+        String eDataType = clonedEvent.edataType();
+        Gson gson = new Gson();
+        Type type = new TypeToken<List<Map<String, String>>>() {
+        }.getType();
+
+        List<Map<String, Object>> items = clonedEvent.edataItems();
+        if (clonedEvent.objectFieldsPresent()) {
+            objectId = clonedEvent.objectID();
+        }
+        for (Map<String, Object> item : items) {
+            clonedEvent.updateEventObjectKey(item.get("id").toString(), item.get("type").toString(), item.get("ver").toString(), objectId);
+            Object itemParams = item.get("params");
+            if (itemParams != null) {
+                List<Map<String, String>> param = gson.fromJson(itemParams.toString(), type);
+                for (Map<String, String> p : param) {
+                    String transfers = p.get("transfers");
+                    if (transfers != null || !transfers.isEmpty()) {
+                        if (Double.parseDouble(transfers) == 0) {
+                            eDataType = DOWNLOAD_KEY;
+                        }
+                        if (Double.parseDouble(transfers) > 0) {
+                            eDataType = IMPORTED_KEY;
+                        }
+                        clonedEvent.updatedEventEdata(eDataType, Double.parseDouble(p.get("size")));
+                    } else {
+                        clonedEvent.updatedEventEdata(orginalEvent.edataType(), null);
+                    }
+                }
+            } else {
+                clonedEvent.updatedEventEdata(orginalEvent.edataType(), null);
```

Comment:
Original object edata attributes "dir" and "type" should also flow into the edata of each flattened event